python/ray/util/spark/cluster_init.py

```diff
@@ -1172,6 +1199,181 @@ def setup_ray_cluster(
+            f"--num-cpus={num_cpus_per_node}",
+            "--block",
+            f"--address={ray_head_ip}:{ray_head_port}",
+            f"--memory={heap_memory_per_node}",
+            f"--object-store-memory={object_store_memory_per_node}",
+            f"--min-worker-port={worker_port_range_begin}",
+            f"--max-worker-port={worker_port_range_end - 1}",
+            f"--dashboard-agent-listen-port={ray_worker_node_dashboard_agent_port}",
+            *_convert_ray_node_options(worker_node_options),
+        ]
+
+        ray_worker_node_extra_envs = {
+            RAY_ON_SPARK_COLLECT_LOG_TO_PATH: collect_log_to_path or "",
+            RAY_ON_SPARK_START_RAY_PARENT_PID: str(os.getpid()),
+            "RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER": "1",
+        }
+
+        if num_gpus_per_node > 0:
+            task_resources = context.resources()
+
+            if "gpu" not in task_resources:
+                raise RuntimeError(
+                    "Couldn't get the gpu id, Please check the GPU resource "
+                    "configuration"
+                )
+            gpu_addr_list = [
+                int(addr.strip()) for addr in task_resources["gpu"].addresses
+            ]
+
+            available_physical_gpus = get_spark_task_assigned_physical_gpus(
+                gpu_addr_list
+            )
+            ray_worker_node_cmd.append(
+                f"--num-gpus={len(available_physical_gpus)}",
+            )
+            ray_worker_node_extra_envs["CUDA_VISIBLE_DEVICES"] = ",".join(
+                [str(gpu_id) for gpu_id in available_physical_gpus]
+            )
+
+        _worker_logger.info(
+            f"Start Ray worker, command: {' '.join(ray_worker_node_cmd)}"
+        )
+
+        try:
+            if autoscale_mode:
+                # Notify job server the task has been launched.
+                requests.post(
+                    url=(
+                        f"http://{ray_head_ip}:{spark_job_server_port}"
+                        "/notify_task_launched"
+                    ),
+                    json={
+                        "spark_job_group_id": spark_job_group_id,
+                    },
+                )
+
+            # Note:
+            # When a pyspark job cancelled, the UDF python worker process are killed by
+            # signal "SIGKILL", then `start_ray_node` process will detect the parent
+            # died event (see `ray.util.spark.start_ray_node.check_parent_alive`) and
+            # then kill ray worker node process and execute cleanup routine.
+            exec_cmd(
+                ray_worker_node_cmd,
+                synchronous=True,
+                extra_env=ray_worker_node_extra_envs,
+            )
+        except Exception as e:
+            if autoscale_mode:
+                _logger.warning(f"Ray worker node process exit, reason: {repr(e)}.")
+            else:
+                raise
+
+        # NB: Not reachable.
+        yield 0
```

Comment:
why `yield 0` instead of `assert False`

Reply:
Because we calls it by `job_rdd.mapPartitions(ray_cluster_job_mapper)`,
spark RDD `mapPartitions` requires the input argument to be a function returns an iterator, or a **python generator**,
so it puts a "yield" inside the function body to convert it to a **python generator**